### PR TITLE
Support top down query

### PR DESCRIPTION
**Pull request by cryptoAtwill**
Add more top down finality util queries in provider.

### CODE REVIEW

File: ipc/provider/src/lib.rs [hunk 1]
```diff
@@ -572,6 +598,25 @@ impl IpcProvider {
...
+    pub async fn get_chain_head_height(&self, subnet: &SubnetID) -> anyhow::Result<ChainEpoch> {
+        let conn = match self.connection(subnet) {
+            None => return Err(anyhow!("target subnet not found")),
+            Some(conn) => conn,
+        };
+
+        conn.manager().chain_head_height().await
+    }
+
+    pub async fn get_genesis_epoch(&self, subnet: &SubnetID) -> anyhow::Result<ChainEpoch> {
```
Comment (adlrocha) [inline ipc/provider/src/lib.rs:611]: Is this a duplicate of the `genesis_epoch` above?

File: ipc/provider/src/lib.rs [hunk 2]
```diff
@@ -572,6 +598,25 @@ impl IpcProvider {
...
+        let parent = subnet.parent().ok_or_else(|| anyhow!("no parent found"))?;
+        let conn = match self.connection(&parent) {
+            None => return Err(anyhow!("parent subnet not found")),
+            Some(conn) => conn,
+        };
+
+        conn.manager().chain_head_height().await
```
Comment (adlrocha) [inline ipc/provider/src/lib.rs:618]: Also, this doesn't seem what you want to do, right? :)

File: ipc/provider/src/lib.rs [hunk 3]
```diff
@@ -534,6 +535,31 @@ impl IpcProvider {
...
+        start: ChainEpoch,
+        end: ChainEpoch,
+    ) -> anyhow::Result<Vec<StakingChangeRequest>> {
+        let conn = match self.connection(subnet) {
+            None => return Err(anyhow!("target subnet not found")),
+            Some(conn) => conn,
+        };
+
+        conn.manager()
+            .get_validator_changeset(subnet, start, end)
```
Comment (adlrocha) [inline ipc/provider/src/lib.rs:559]: Where is this one implemented?

File: ipc/provider/src/manager/subnet.rs [hunk 4]
```diff
@@ -128,6 +128,8 @@ pub trait SubnetManager: Send + Sync + TopDownCheckpointQuery {
 /// Trait to interact with a subnet to query the necessary information for top down checkpoint.
 #[async_trait]
 pub trait TopDownCheckpointQuery: Send + Sync {
+    /// Returns the chain head height
```
Comment (adlrocha) [inline ipc/provider/src/manager/subnet.rs:131]: This comment is wrong, right?

File: ipc/provider/src/lib.rs [hunk 5]
```diff
@@ -572,6 +598,25 @@ impl IpcProvider {
 
         conn.manager().get_chain_id().await
     }
+
+    pub async fn get_chain_head_height(&self, subnet: &SubnetID) -> anyhow::Result<ChainEpoch> {
+        let conn = match self.connection(subnet) {
```
Comment (adlrocha) [inline ipc/provider/src/lib.rs:603]: Isn't the change set picked up from the parent?

**Review by adlrocha — CHANGES_REQUESTED** [review]
There seems to be a lot of copy-pasta and things missing (or maybe I am looking at the wrong commit?)

File: ipc/provider/src/lib.rs (hunk 5 above)
Comment (cryptoAtwill) [reply ipc/provider/src/lib.rs:603]: this is get chain head height? Anyways, for change set, it's from the subnet actor in parent, so the address should be derived from the subnet directly.

File: ipc/provider/src/lib.rs (hunk 2 above)
Comment (cryptoAtwill) [reply ipc/provider/src/lib.rs:618]: right, this method should be removed.

**Review by adlrocha — APPROVED** [review]

**Pull request merged**